File: Ricart.java
import java.util.*;

class Request {
    int site;
    int timestamp;

    public Request(int site, int timestamp) {
        this.site = site;
        this.timestamp = timestamp;
    }
}

public class Ricart {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the total number of processes (sites): ");
        int totalSites = sc.nextInt();

        System.out.print("Enter the number of sites requesting to enter the critical section: ");
        int requestingSites = sc.nextInt();

        int[] timestamps = new int[totalSites + 1];  // site index from 1
        Request[] requests = new Request[requestingSites];
        ArrayList<Integer>[] deferredReplies = new ArrayList[totalSites + 1];

        // Initialize deferredReplies lists
        for (int i = 1; i <= totalSites; i++) {
            deferredReplies[i] = new ArrayList<>();
        }

        // Read timestamps for all sites
        for (int i = 1; i <= totalSites; i++) {
            System.out.print("Enter the timestamp for Site " + i + ": ");
            timestamps[i] = sc.nextInt();
        }

        // Read which sites want to enter critical section
        for (int i = 0; i < requestingSites; i++) {
            System.out.print("Enter the site number that wants to enter CS: ");
            int site = sc.nextInt();
            requests[i] = new Request(site, timestamps[site]);
        }

        // Handle reply/defer logic
        for (Request r : requests) {
            System.out.println("\n--- Site " + r.site + " wants to enter CS ---");
            for (int j = 1; j <= totalSites; j++) {
                if (j == r.site) continue;

                if (timestamps[j] <= r.timestamp) {
                    System.out.println("Site " + j + " deferred reply to Site " + r.site);
                    deferredReplies[j].add(r.site);
                }else{
                    System.out.println("Site " + j + " replied to Site " + r.site);
                }
            }
        }

        // Sort the requests based on timestamp
        Arrays.sort(requests, Comparator.comparingInt(r -> r.timestamp));

        // Simulate entering critical section
        System.out.println("\n--- Critical Section Access ---");
        for (Request r : requests) {
            System.out.println("Site " + r.site + " entered CS");

            for (int deferredSite : deferredReplies[r.site]) {
                System.out.println("Site " + r.site + " replies to deferred Site " + deferredSite);
            }

            deferredReplies[r.site].clear(); // After replying
        }

        sc.close();
    }
}

File: DC LAB/DC LAB/7 RICART/ricart.c
#include <stdio.h>
#include <stdlib.h>
void main()
{
    int ns, ncs, timestamp, site;
    printf("Enter number of sites: ");
    scanf("%d", &ns);
    printf("Enter number of sites which want to enter critical section: ");
    scanf("%d", &ncs);
    int ts[ns], request[ncs], mp_timestamp[ncs], mp_site[ncs], deferred[ns];
    for (int i = 0; i < ns; i++)
    {
        ts[i] = 0;
        deferred[i] = 0;
    }
    for (int i = 0; i < ncs; i++)
    {
        printf("\nEnter timestamp: ");
        scanf("%d", &timestamp);
        printf("Enter site number: ");
        scanf("%d", &site);
        ts[site - 1] = timestamp;
        request[i] = site;
        mp_timestamp[i] = timestamp;
        mp_site[i] = site;
    }
    printf("\nSites and Timestamp:\n");
    for (int i = 0; i < ns; i++)
        printf("%d %d\n", i + 1, ts[i]);
    for (int i = 0; i < ncs; i++)
    {
        printf("\nRequest from site: %d\n", request[i]);
        for (int j = 0; j < ns; j++)
        {
            if (request[i] != (j + 1))
            {
                if (ts[j] > ts[request[i] - 1] || ts[j] == 0)
                {
                    printf("%d Replied\n", j + 1);
                }
                else
                {
                    printf("%d Deferred\n", j + 1);
                    deferred[j] = request[i];
                }
            }
        }
    }
    for (int i = 0; i < ncs - 1; i++)
    {
        for (int j = 0; j < ncs - i - 1; j++)
        {
            if (mp_timestamp[j] > mp_timestamp[j + 1])
            {
                int temp_time = mp_timestamp[j];

                mp_timestamp[j] = mp_timestamp[j + 1];
                mp_timestamp[j + 1] = temp_time;
                int temp_site = mp_site[j];
                mp_site[j] = mp_site[j + 1];
                mp_site[j + 1] = temp_site;
            }
        }
    }
    printf("\nSites entered Critical Section:\n");
    for (int i = 0; i < ncs; i++)
    {
        printf("Site %d entered Critical Section\n", mp_site[i]);
        if (deferred[mp_site[i] - 1])
        {
            printf("Site %d then checks its deferred array and replies to Site %d\n", mp_site[i],
                   deferred[mp_site[i] - 1]);
            deferred[mp_site[i] - 1] = 0;
        }
    }
}